### PR TITLE
PMM-3074 Draft API for MySQL and PostgreSQL.

### PR DESCRIPTION
**Pull request by BupycHuk**

### CODE REVIEW

File: api/mysql.proto [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+message MySQLRemoveRequest {
+    MySQLInstanceID id = 1;
+}
+
+message MySQLRemoveResponse {
+}
+
+service MySQL {
+    rpc Discover(MySQLDiscoverRequest) returns (MySQLDiscoverResponse) {
```
Comment (AlekSi) [inline api/mysql.proto]: We will not have that API in 1.16, let's drop it.

File: api/mysql.proto [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        option (google.api.http) = {
+            post: "/v0/mysql/discover"
+            body: "*"
+        };
+    }
+
+    rpc List(MySQLListRequest) returns (MySQLListResponse) {
+        option (google.api.http) = {
+            get: "/v0/mysql"
```
Comment (AlekSi) [inline api/mysql.proto]: Let's use POST with a body for all new APIs, and not try to emulate REST. I will explain why in Slack.

So, something like `/v0/mysql/list`.

File: api/mysql.proto [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    rpc Add(MySQLAddRequest) returns (MySQLAddResponse) {
+        option (google.api.http) = {
+            post: "/v0/mysql"
+            body: "*"
+        };
+    }
+
+    rpc Remove(MySQLRemoveRequest) returns (MySQLRemoveResponse) {
+        option (google.api.http) = {
+            delete: "/v0/mysql"
```
Comment (AlekSi) [inline api/mysql.proto]: POST with a body. `/v0/mysql/remove`.

File: api/mysql.proto [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    rpc List(MySQLListRequest) returns (MySQLListResponse) {
+        option (google.api.http) = {
+            get: "/v0/mysql"
+        };
+    }
+
+    rpc Add(MySQLAddRequest) returns (MySQLAddResponse) {
+        option (google.api.http) = {
+            post: "/v0/mysql"
```
Comment (AlekSi) [inline api/mysql.proto]: `/v0/mysql/add`

File: api/postgresql.proto [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        };
+    }
+
+    rpc Remove(PostgreSQLRemoveRequest) returns (PostgreSQLRemoveResponse) {
+        option (google.api.http) = {
+            delete: "/v0/postgresql"
+            body: "*"
+        };
+    }
+}
```
Comment (AlekSi) [inline api/postgresql.proto]: Please do the same as for MySQL.

**Comment by AlekSi** [comment]
Please click the button at https://github.com/percona/pmm-managed/pull/61#issuecomment-428146450.

File: api/mysql.proto [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+message MySQLListRequest {
+}
+
+message MySQLListResponse {
+    repeated MySQLInstance instances = 1;
+}
+
+message MySQLAddRequest {
+    MySQLInstanceID id = 1;
```
Comment (AlekSi) [inline api/mysql.proto]: We will need host and port, not region and name.

File: api/mysql.proto [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    string address = 4;
+    uint32 port = 5;
+    string engine = 6;
+    string engine_version = 7;
+}
+
+message MySQLInstanceID {
+    string region = 1;
+    string name = 2;
+}
```
Comment (AlekSi) [inline api/mysql.proto]: That's RDS-specific.

File: api/mysql.proto [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+package api;
+
+import "google/api/annotations.proto";
+
+message MySQLNode {
+    reserved 1, 2; // id and type
+
+    string region = 3;
+    string name = 4;
```
Comment (AlekSi) [inline api/mysql.proto]: What are the region and name there?

File: api/mysql.proto [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+message MySQLRemoveRequest {
+    MySQLInstanceID id = 1;
+}
+
+message MySQLRemoveResponse {
+}
+
+service MySQL {
+
```
Comment (AlekSi) [inline api/mysql.proto]: Remove empty line.

File: api/mysql.proto (hunk 8 above)
Comment (BupycHuk) [reply api/mysql.proto]: we have host and port in `MySQLService`, should we duplicate it here?

File: api/mysql.proto (hunk 7 above)
Comment (BupycHuk) [reply api/mysql.proto]: got it

File: api/mysql.proto [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+message MySQLAddRequest {
+    string address = 1;
+    uint32 port = 2;
+    string username = 3;
+    string password = 4;
+}
+
+message MySQLAddResponse {
+}
```
Comment (AlekSi) [inline api/mysql.proto]: Let's return id.

File: api/mysql.proto [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+message MySQLRemoveRequest {
+    string address = 1;
+    uint32 port = 2;
```
Comment (AlekSi) [inline api/mysql.proto]: Let's use id instead.

File: api/mysql.proto (hunk 9 above)
Comment (AlekSi) [reply api/mysql.proto]: ^

File: api/mysql.proto [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        option (google.api.http) = {
+            post: "/v0/mysql"
+            body: "*"
+        };
+    }
+
+    rpc Remove(MySQLRemoveRequest) returns (MySQLRemoveResponse) {
+        option (google.api.http) = {
+            delete: "/v0/mysql"
+            body: "*"
```
Comment (AlekSi) [inline api/mysql.proto]: Let's add id to the URL and drop body. Generator support that natively.

File: api/mysql.proto (hunk 11 above)
Comment (BupycHuk) [reply api/mysql.proto]: ok

File: api/mysql.proto (hunk 10 above)
Comment (BupycHuk) [reply api/mysql.proto]: ok

File: api/mysql.proto (hunk 12 above)
Comment (BupycHuk) [reply api/mysql.proto]: ok

**Pull request merged**